autoria/api.py

```diff
@@ -277,148 +277,245 @@ class RiaAverageCarPrice:
+        state = None
+        city = None
+        for key in kwargs:
+            # To shorten the code we call the setter method by name,
+            # this is same as if we did this:
+            #
+            # for key in kwargs:
+            #     if key == 'state':
+            #         self.state(kwargs['state'])
+            #     if key == 'bodystyle':
+            #         self.bodystyle(kwargs['bodystyle'])
+            #      ... and so on
+            if not hasattr(self, key):
+                raise Exception('Unknown parameter: {}'.format(key))
+            # We have special handling for state and city as we need to
+            # process them together.
+            if key == 'state':
+                state = kwargs['state']
+            elif key == 'city':
+                city = kwargs['city']
+            else:
+                getattr(self, key)(kwargs[key])
+
+        if state and city:
+            self.city(state, city)
+        elif state:
+            self.state(state)
+
+    def get_average(self) -> dict:
+        """Get average price for composed search parameters."""
+        return self._api.average_price(self.as_dict())
+
+    def _category(self, category):
+        """Set category.
+
+        The method is private because category is a required constructor
+        parameter.
+        """
+        self._category_id = select_item(category, self._api.get_categories())
+
+    def _mark(self, mark):
+        """Set mark.
+
+        The method is private because mark is a required constructor
+        parameter.
+        """
+        self._mark_id = select_item(
+            mark, self._api.get_marks(self._category_id))
+
+    def _model(self, model):
+        """Set model.
+
+        The method is private because model is a required constructor
+        parameter.
+        """
+        self._model_id = select_item(
             model,
-            self._api.get_models(category_id, mark_id)
+            self._api.get_models(self._category_id, self._mark_id)
         )
 
-        # Processing the rest of args, those which are defaulted to None
-        # are processed below
-        self._params = {
-            'main_category': category_id,
-            'marka_id': mark_id,
-            'model_id': model_id,
-            'state_id': None,
-            'body_id': None,
-            'city_id': None,
-            'yers': years,
-            'raceInt': mileage,
-            'gear_id': None,
-            'options': None,
-            'fuel_id': None,
-            'drive_id': None,
-            'color_id': None,
-            'engineVolume': engine_volume,
-            'seats': seats,
-            'door': doors,
-            'carrying': carrying,
-            'custom': custom,
-            'damage': damage,
-            'under_credit': under_credit,
-            'confiscated_car': confiscated,
-            'onRepairParts': on_repair_parts,
-        }
+    def bodystyle(self, bodystyle):
```

Comment:
I think I am missing some setters for plain variables - years, mileage and engine_volume, maybe something else.